### PR TITLE
[Merged by Bors] - Move malfeasance proofs types to malfeasance/wire pkg

### PR DESCRIPTION
**Pull request by poszu**
## Motivation

To avoid cyclic dependencies, required to split out an ATX wire type.

## Description

- moved `common/types/malfeasance*.go` to `malfeasance/wire` pkg

## Test Plan

n/a

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [ ] Update documentation as needed
- [ ] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

File: api/grpcserver/mesh_service.go [hunk 1]
```diff
@@ -594,7 +595,7 @@ func (s MeshService) MalfeasanceStream(
 	}
 
 	// first serve those already existed locally.
-	if err := s.cdb.IterateMalfeasanceProofs(func(id types.NodeID, mp *types.MalfeasanceProof) error {
+	if err := s.cdb.IterateMalfeasanceProofs(func(id types.NodeID, mp *wire.MalfeasanceProof) error {
```
Comment (fasmat) [inline api/grpcserver/mesh_service.go:598]: Probably for next PR.

There might also need to be a distinction between the wire-type `MalfeasanceProof` and the domain type of it. Ideally sync/gossip only receive the bytes and call the handler in the `malfeasance` package which itself then calls the handlers for the different types of proofs in their respective packages (via interfaces).

This way `malfeasance` is a dependency of (most) other packages while it itself doesn't depend on anything else.

File: events/events.go [hunk 2]
```diff
@@ -283,26 +284,26 @@ func emitUserEvent(help string, failure bool, details pb.IsEventDetails) {
 	}
 }
 
-func ToMalfeasancePB(nodeID types.NodeID, mp *types.MalfeasanceProof, includeProof bool) *pb.MalfeasanceProof {
+func ToMalfeasancePB(nodeID types.NodeID, mp *wire.MalfeasanceProof, includeProof bool) *pb.MalfeasanceProof {
```
Comment (fasmat) [inline events/events.go:287]: This probably also needs to be in the `malfeasance` package and call handlers in other packages (via interfaces) to convert a typed malfeasance proof into the GRPC type that we then return via the event stream.

**Review by fasmat — APPROVED** [review]

**Comment by poszu** [comment]
bors merge

File: api/grpcserver/mesh_service.go (hunk 1 above)
Comment (poszu) [reply api/grpcserver/mesh_service.go:598]: Maybe the `malfeasance` pkg could export a function to create malfeasance proofs of a given type, and with given content so others don't need to depend on `malfeasance/wire`. Something like:

```go
package malfeasance

type ProofKind byte

const (
	MultipleATXs ProofKind = iota + 1
	MultipleBallots
	HareEquivocation
	InvalidPostIndex
)

type ProofContent interface {
	scale.Type
}

type MalfeasanceProof interface {
	... methods needed, probably a way to serialize for publish?
}

func CreateProof(layer types.LayerID, kind ProofKind, content ProofContent) MalfeasanceProof {
	...
}
```

The malfeasance Handler could have a set of validators registered in the form of a map:
```go
type Validator interface {
	Validate(data []byte) (*types.NodeID, ProofContent, error)
}

validators map[ProofKind]Validator
```

Then, the wire-type for, for example, ATX equivocation (`MultipleATXs`) would be in `activation` package along with a validator for it.

**Comment by poszu** [comment]
bors merge

**Pull request closed without merging**